dev/domain-controller/new-agent-vm/Install-PCoIPAgent.sh

```diff
@@ -0,0 +1,210 @@
+#!/bin/bash
+
+install_driver()
+{
+    # Install kernel-source/kernel-devel and gcc
+    sudo yum -y install kernel-source kernel-devel gcc
+    FILE_NAME='NVIDIA-Linux-x86_64-367.106-grid.run'
+    FILE_LOCATION='/root/'$FILE_NAME
+    # Download Driver first
+    echo "Downloading and Installing Nvidia driver"
+    sudo wget --retry-connrefused --tries=3 --waitretry=5 -O $FILE_LOCATION https://binarystore.blob.core.windows.net/thirdparty/nvidia/$FILE_NAME
+    exitCode=$?
+    if [ $exitCode -ne 0 ]
+    then
+        echo "failed to download Nvidia driver."
+        # let's define exit code 103 for this case
+        exit 103
+    fi
+    # Change file permission
+    sudo chmod 744 $FILE_LOCATION
+    # run installer
+    sudo $FILE_LOCATION -Z -X -s
+    exitCode=$?
+    
+    if [ $exitCode -eq 0 ]
+    then
+        echo "Driver is installed successfully"
+    else
+        echo "failed to install Nvidia driver. Will create a script to install driver when machine boots up"
+        file_path=/root/install_driver.sh
+        cat <<EOF >$file_path
+#!/bin/bash
+if [ -e /root/.first_boot ]
+then
+    if [ -e /root/.second_boot ]
+    then
+        exit 0
+    else
+        touch /root/.second_boot
+        sudo $FILE_LOCATION -Z -X -s        
+    fi
+else
+    touch /root/.first_boot
+    sudo $FILE_LOCATION -Z -X -s
+    (sleep 2;  sudo shutdown -f -r +0)&
+fi
+EOF
+        sudo chmod 744 $file_path
+        (crontab -l 2>/dev/null; echo "@reboot $file_path") | crontab -
+        # let's define exit code 104 for this case
+        #exit 104
+    fi
+}
+
+# the first argument is the Registration Code of PCoIP agent
+REGISTRATION_CODE=$1
+# the second argument is the agent type
+AGENT_TYPE=$2
+# the third argument is the VM name
+VM_NAME=$3
+# the forth argument is the domain name
+DOMAIN_NAME=$4
+# the fifith argument is the username
+USERNAME=$5
+# the sixth argument is the password
+PASSWORD=$6
+# the seventh argument is the domain group to join
+GROUP=$7
+
+# Make sure Linux OS is up to date
+echo "--> Updating Linux OS to latest"
+# Exclude WALinuxAgent due to it failing to update from within an Azure Custom Script
+sudo yum -y update --exclude=WALinuxAgent
+
+# If it's graphic agent, install Nvidia Driver
+case "$AGENT_TYPE" in 
+    "Graphics")
+        install_driver
+        AGENT_TYPE='graphics'
+        ;;
+    "Standard")
+        AGENT_TYPE='standard'
+        ;;   
+    *)
+        echo "unknown agent type $AGENT_TYPE."
+        # let's define exit code 105 for this case
+        exit 105
+        ;;       
+esac
+
+# Join domain
+echo "-->Install required packages to join domain"
+sudo yum -y install sssd realmd oddjob oddjob-mkhomedir adcli samba-common samba-common-tools krb5-workstation openldap-clients policycoreutils-python
+sudo systemctl enable sssd
+
+echo "-->Joining the domain"
```

Comment:
Could this be a function? And only call it if all Domain Credentials are provided (otherwise don't join domain)?